lib/internal/modules/esm/loader.js

```diff
@@ -225,6 +231,39 @@ class ESMLoader {
     };
   }
 
+  /**
+   * Returns the url to use for the resolution of a given cache key url
+   * These are not guaranteed to be the same.
+   *
+   * In WHATWG HTTP spec for ESM the cache key is the non-I/O bound
+   * synchronous resolution using only string operations
+   *   ~= resolveImportMap(new URL(specifier, importerHREF))
+   *
+   * The url used for subsequent resolution is the response URL after
+   * all redirects have been resolved.
+   *
+   * https://example.com/foo redirecting to https://example.com/bar
+   * would have a cache key of https://example.com/foo and baseURL
+   * of https://example.com/bar
+   *
+   * MUST BE SYNCHRONOUS for import.meta initialization
+   * MUST BE CALLED AFTER receiving the url body due to I/O
+   * @param {string} url
+   * @returns {string}
```

Comment:
```suggestion
   * @param {string} url The URL specifier to look up
   * @returns {URL['href']} The resolved url for the provided specifier
```

Reply:
like the `URL['href']` trick, haven't used that before